core/src/main/java/kafka/server/ClientMetricsManager.java

```diff
@@ -34,13 +70,348 @@ public class ClientMetricsManager implements Closeable {
+        } catch (ApiException exception) {
+            clientInstance.lastKnownError(Errors.forException(exception));
+            return request.getErrorResponse(throttleMs, exception);
+        } finally {
+            // Update the client instance with the latest push request parameters.
+            clientInstance.terminating(request.data().terminating());
+            clientInstance.lastPushRequestEpoch(now);
+        }
+
+        // Push the metrics to the external client receiver plugin.
+        byte[] metrics = request.data().metrics();
+        if (metrics != null && metrics.length > 0) {
+            ClientMetricsReceiverPlugin.exportMetrics(requestContext, request);
+        }
+
+        clientInstance.lastKnownError(Errors.NONE);
+        return request.createResponse(throttleMs, Errors.NONE);
     }
 
     @Override
     public void close() throws IOException {
-        // TODO: Implement the close logic to close the client metrics manager.
+        // Do nothing for now.
+    }
+
+    private void updateLastSubscriptionUpdateEpoch() {
+        this.lastSubscriptionUpdateEpoch = System.currentTimeMillis();
+    }
+
+    private void updateClientSubscription(String subscriptionName, ClientMetricsConfigs configs) {
+        List<String> metrics = configs.getList(ClientMetricsConfigs.SUBSCRIPTION_METRICS);
+        int pushInterval = configs.getInt(ClientMetricsConfigs.PUSH_INTERVAL_MS);
+        List<String> clientMatchPattern = configs.getList(ClientMetricsConfigs.CLIENT_MATCH_PATTERN);
+
+        SubscriptionInfo newSubscription =
+            new SubscriptionInfo(subscriptionName, metrics, pushInterval,
+                ClientMetricsConfigs.parseMatchingPatterns(clientMatchPattern));
+
+        subscriptionMap.put(subscriptionName, newSubscription);
+    }
+
+    private Uuid generateNewClientId() {
+        Uuid id = Uuid.randomUuid();
+        while (clientInstanceCache.get(id) != null) {
+            id = Uuid.randomUuid();
+        }
+        return id;
+    }
+
+    private ClientMetricsInstance getClientInstance(Uuid clientInstanceId, RequestContext requestContext,
+        long timestamp) {
+        // Check if null can be called on the cache. if can then we can avoid the method call.
+        ClientMetricsInstance clientInstance = clientInstanceCache.get(clientInstanceId);
+
+        if (clientInstance == null) {
+            // If the client instance is not present in the cache, then create a new client instance
+            // and update the cache. This can also happen when the telemetry request is received by
+            // the separate broker instance.
+            ClientMetricsInstanceMetadata instanceMetadata = new ClientMetricsInstanceMetadata(
+                clientInstanceId, requestContext);
+            clientInstance = createClientInstanceAndUpdateCache(clientInstanceId, instanceMetadata, timestamp);
+        } else if (clientInstance.subscriptionUpdateEpoch() < lastSubscriptionUpdateEpoch) {
```

Comment:
As I do not think that subscription creation/updation would be too frequent i.e. number of subscriptions should be fairly low in broker hence the situation where client instance is created at the same time of subscription should be rare. But when a broker is re-started and there are clients waiting for the connection, then subscription ll be loaded (from dynamic config handler) at the same time when instance is created hence I skipped `=` check to avoid re-evaluation of subscriptions.

 Also in the next telemetry API call re-evaluation of update timestamp will happen which shall either update the subscription or throw appropriate error to client.

Reply:
Yes you are right, that could be a problem. But using atomic integer can bump the version appropriately but we need a definite way to track the subscription evaluation version in client instance. Hence either we need to have read and write locks to store correct subscription version in client instance or, the approach I took in the PR, store the version with each subscription info and while evaluating subscriptions for client instance, store the maximum version encountered.

I have addressed the changes required, done.

Reply:
Moved to single global variable, resolving the comment.